File: media/lua/shared/module.lua
require 'asledgehammer/class'

Module = class(function(o, id, name)

    -- Set the variables.
    o.id = id;
    o.name = name;
    o.version = '1.00';
    o.loaded = false;
    o.started = false;
    o.handshaked = false;
end);

function Module:sendCommand(command, args)
    -- Sends the command to CraftHammer.
    CraftHammer.instance:sendCommand('crafthammer.module.' .. self:getID(),
                                     command, args);
end

function Module:sendRequest(command, args, success, failure)
    local request = Request('crafthammer.module.' .. self:getID(), command,
                            args, success, failure);
    request:send();
end

----------------------------------------------------------------
--
----------------------------------------------------------------
function Module:writeFile(args)
    local packet = args.packet;
    local execute = false;
    if packet == 0 then
        self._explodedFile = args.explodedFile;
        self._explodedFile.segment = 0;
        self._explodedFile.fileData = {};
        self._explodedFile.path =
            "CraftHammer/module/" .. self:getName() .. "/" ..
                self._explodedFile.path;
    elseif packet == 2 then
        execute = true;
    end
    self._explodedFile.segment = self._explodedFile.segment + 1;
    self._explodedFile.fileData[tLength(self._explodedFile.fileData)] =
        args.data;
    print("(" .. self._explodedFile.path .. ") " ..
              tostring(self._explodedFile.segment) .. "/" ..
              tostring(self._explodedFile.segments));
    if execute then
        writeFile(self._explodedFile);
        self._explodedFile = nil;
    end
end

----------------------------------------------------------------
-- @return 	Returns the version of the Module.
----------------------------------------------------------------
function Module:getVersion() return self.version; end

----------------------------------------------------------------
-- Loads the module. Create objects here.
----------------------------------------------------------------
function Module:load() end

----------------------------------------------------------------
-- Starts the module. Add listeners here.
----------------------------------------------------------------
function Module:start() end

----------------------------------------------------------------
-- Place initial Command requests here.
----------------------------------------------------------------
function Module:handshake() end

----------------------------------------------------------------
-- On every in-game Tick, this function is executed.
----------------------------------------------------------------
function Module:update() end

----------------------------------------------------------------
-- Handles a command from the server.
--
-- @string command 	The command given from the server.
-- @table args 		The arguments given with the command.
----------------------------------------------------------------
function Module:command(command, args) end

----------------------------------------------------------------
-- Stops the module. Remove listeners here.
----------------------------------------------------------------
function Module:stop() end

----------------------------------------------------------------
-- Unloads the module. Nullify or deconstruct objects here.
----------------------------------------------------------------
function Module:unload() end

----------------------------------------------------------------
-- @return 	Returns the name of the Module.
----------------------------------------------------------------
function Module:getName() return self.name; end

----------------------------------------------------------------
-- @return 	Returns the ID of the Module.
----------------------------------------------------------------
function Module:getID() return self.id; end

----------------------------------------------------------------
-- @return 	Returns if the Module has started.
----------------------------------------------------------------
function Module:isStarted() return self.started; end

----------------------------------------------------------------
-- @return 	Returns if the Module has handshaked.
----------------------------------------------------------------
function Module:isHandshaked() return self.handshaked; end

----------------------------------------------------------------
-- @return 	Returns if the Module is currently stopped.
----------------------------------------------------------------
function Module:isStopped() return not self.started; end

----------------------------------------------------------------
-- @return 	Returns if the Module has been loaded.
----------------------------------------------------------------
function Module:isLoaded() return self.loaded; end

----------------------------------------------------------------
-- @return 	Returns if the Module is not loaded.
----------------------------------------------------------------
function Module:isUnloaded() return not self.loaded; end

File: src/shared/crafthammer/module.ts
import { Request, RequestError, RequestSuccess } from './util/request';

export abstract class Module {
    readonly properties: ModuleProperties;

    private _loaded: boolean = false;
    private _started: boolean = false;
    private _handshaked: boolean = false;

    constructor(properties: ModuleProperties) {
        this.properties = properties;
    }

    sendRequest(command: string, args: any[], timeout: number, success: RequestSuccess, error: RequestError) {
        let request = new Request('crafthammer.module.' + this.properties.id, command, args, timeout, success, error);
    }

    public get loaded(): boolean {
        return this._loaded;
    }

    public get started(): boolean {
        return this._started;
    }

    public get handshaked(): boolean {
        return this._handshaked;
    }

    public handshake() {
        this.onHandshake();
        this._handshaked = true;
    }

    public load() {
        const { id } = this.properties;
        if (this._loaded) {
            throw new Error(`The module is already loaded: ${id}`);
        }
        this.onLoad();
        this._loaded = true;
    }

    public start() {
        const { id } = this.properties;
        if (!this._loaded) {
            throw new Error(`The module is not loaded: ${id}`);
        } else if (this._started) {
            throw new Error(`The module has already started: ${id}`);
        }
        this.onStart();
        this._started = true;
    }

    public update() {
        const { id } = this.properties;
        if (!this._loaded) {
            throw new Error(`The module is not loaded: ${id}`);
        } else if (!this._started) {
            throw new Error(`The module has not started: ${id}`);
        }
        this.onUpdate();
    }

    public stop() {
        const { id } = this.properties;
        if (!this._loaded) {
            throw new Error(`The module is not loaded: ${id}`);
        } else if (!this._started) {
            throw new Error(`The module has not started: ${id}`);
        }
        this.onStop();
        this._started = false;
    }

    public unload() {
        const { id } = this.properties;
        if (!this._loaded) {
            throw new Error(`The module is not loaded: ${id}`);
        } else if (this._started) {
            throw new Error(`The module is running and cannot be unloaded: ${id}`);
        }
        this.onUnload();
        this._loaded = false;
    }

    protected onHandshake() {}
    protected onLoad() {}
    protected onStart() {}
    protected onUpdate() {}
    protected onStop() {}
    protected onUnload() {}
}

export type ModuleProperties = {
    id: string;
    name: string;
    version: string;
};
